[Namba1팀]backend ProductDetailView

nicholas019 opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [ ] 리뷰 반영
- [ ] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- ProductDetailView 구현

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1. path parameter를 사용하여 상세페이지 조회기능 구현

<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
- 엔드포인트에 path parameter 사용법의 이해

<br />

## :: 기타 질문 및 특이 사항

nicholas019 committed:
[Namba1팀]backend ProductDetaillView 수정1

nicholas019 committed:
[Namba1팀]backend ProductDetailView 수정1

nicholas019 committed:
[Namba1팀]backend ProductDetailView 수정2

nicholas019 committed:
[Namba1팀]backend ProductDetailView 수정3

Jon9Beom commented on products/views.py:
```diff
@@ -38,4 +38,28 @@ def get(self, request):
+    def get(self, request, id):
+        try:
+            products = Product.objects.filter(id = id)
+            product_detail = [{
+                "id"         : product.id,
+                "name"       : product.name,
+                "description": product.description,
+                "serving"    : product.serving,
+                "cookTime"   : product.cook_time,
+                "prepTime"   : product.prep_time,
+                "price"      : int(product.price),
+                "spice"      : [spice.level for spice in product.spice_set.all()],
+                "image"      : [image.image for image in product.productimage_set.all()]
+            } for product in products]
```
product_detail을 리스트 컴프리헨션으로 작성하신 이유가 있으신가요-?
바로 윗줄에서 오브젝트를 id(pk)값으로 필터를 한다면, 상품 디테일 API의 역할 그대로 하나의 오브젝트 만을 가져오기 때문에 굳이 리스트 컴프리헨션으로 작성하지 않으셔도 될 것 같습니다.
```suggestion
            product_detail = {
                "id"         : product.id,
                "name"       : product.name,
                "description": product.description,
                "serving"    : product.serving,
                "cookTime"   : product.cook_time,
                "prepTime"   : product.prep_time,
                "price"      : int(product.price),
                "spice"      : [spice.level for spice in product.spice_set.all()],
                "image"      : [image.image for image in product.productimage_set.all()]
            } 
```

nicholas019 replied:
네 바로 수정했습니다 그리고 postman으로 해보니 오류없이 통과되었습니다 어차피 하나의값만 받으니 당연히 get 메소드로 사용해야되는데 놓첬습니다 감사합니다

Jon9Beom commented on products/views.py:
```diff
@@ -38,4 +38,28 @@ def get(self, request):
+
+        except KeyError:
+            return JsonResponse({"message": 'KeyError'}, status = 400)  
+        except Product.DoesNotExist:
+            return JsonResponse({"message": 'No Data'}, status = 400)
```
django의 orm 중 filter와 get,
이 두가지 메소드는 싱글 오브젝트 즉, 하나의 쿼리셋만 불러오더라도 차이가 있습니다.

get의 경우에는 해당하는 오브젝트가 없을 때 DoesNotExist 에러가 발생하지만 filter의 경우에는 단지 빈 쿼리셋을 반환할 뿐 에러가 발생하진 않습니다.

수정해야할 사항이라기보다는 짚고 넘어가야 할 내용이라 코멘트 남겨드렸습니다!

nicholas019 replied:
네 감사합니다 위의 피드백대로 get으로 메소드로 변경하여 DoesNotExist 에러는 그대로 두었습니다 감사합니다

nicholas019 committed:
[NAMBA1] backend ProductDetailView 수정

수정사항
1. 47. filter > get 변경
2. 수정1번으로 인해 for문으로 product_detail을 할필요가없어 list comprehension 삭제

Jon9Beom commented:
LGTM 🚀
